src/pe.cc

```diff
@@ -0,0 +1,280 @@
+
+  if (dos_header_.Magic != dos_magic) {
+    // Not a PE file.
+    return false;
+  }
+
+  if ((dos_header_.AddressOfNewExeHeader + sizeof(pe_header)) > data_.size()) {
+    // Cannot fit the headers
+    return false;
+  }
+
+  memcpy(&pe_header_, data_.data() + dos_header_.AddressOfNewExeHeader,
+         sizeof(pe_header_));
+
+  if (!std::equal(pe_header_.signature, pe_header_.signature + sizeof(PE_Magic),
+                  std::begin(PE_Magic))) {
+    // Not a PE file.
+    return false;
+  }
+
+  // TODO(mj): Parse PE header further to determine this
+  is_64bit_ = false;
+
+  section_count_ = pe_header_.NumberOfSections;
+
+  const uint32_t sections_offset = dos_header_.AddressOfNewExeHeader +
+                                   sizeof(pe_header) +
+                                   pe_header_.SizeOfOptionalHeader;
+
+  auto sections_size = CheckedMul(section_count_, sizeof(pe_section));
+  if ((sections_offset + sections_size) > data_.size()) {
+    // Cannot fit the headers
+    return false;
+  }
+
+  header_region_ = GetRegion(0, sections_offset);
+  section_headers_ = GetRegion(sections_offset, sections_size);
+
+  return true;
+}
+
+class Section {
+ public:
+  std::string name;
+  string_view data;
+
+  uint32_t raw_offset() const { return header_.PointerToRawData; }
+  uint32_t raw_size() const { return header_.SizeOfRawData; }
+
+  uint32_t virtual_addr() const { return header_.VirtualAddress; }
+  uint32_t virtual_size() const { return header_.VirtualSize; }
+
+  Section(string_view header_data) {
+    assert(header_data.size() == sizeof(header_));
+    memcpy(&header_, header_data.data(), sizeof(header_));
+    data = header_data;
+
+    // TODO(mj): Handle long section names:
+    // For longer names, this member contains a forward slash (/) followed by an
+    // ASCII representation of a decimal number that is an offset into the
+    // string table.
+    name = std::string(header_.Name,
+                       strnlen(header_.Name, STRUCT_SIZES::NameSize));
+  }
+
+ private:
+  pe_section header_;
+};
+
+template <class Func>
+void ForEachSection(const PeFile& pe, Func&& section_func) {
+  for (auto n = 0; n < pe.section_count(); ++n) {
+    Section section(pe.section_header(n));
+    section_func(section);
+  }
+}
+
+void ParseSections(const PeFile& pe, RangeSink* sink) {
+  assert(pe.IsOpen());
+  ForEachSection(pe, [sink](const Section& section) {
+    uint64_t vmaddr = section.virtual_addr();
+    uint64_t vmsize = section.virtual_size();
+
+    uint64_t fileoff = section.raw_offset();
+    uint64_t filesize = section.raw_size();
+
+    sink->AddRange("pe_sections", section.name, vmaddr, vmsize, fileoff,
+                   filesize);
+  });
+}
+
+void AddCatchAll(const PeFile& pe, RangeSink* sink) {
+  // The last-line fallback to make sure we cover the entire VM space.
+  assert(pe.IsOpen());
+
+  auto begin = pe.header_region().data() - sink->input_file().data().data();
```

Comment:
Had to fix this line, since absl uses std::string_view with c++17,
previously this was:
`auto begin = pe.header_region().begin() - sink->input_file().data().begin()`

However msvc iterators asserted on this, complaining that they were not compatible.
Might be something that could also be used to absl string_view iterators?